dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/EndpointDefinitions.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.annotations.processor.data;
+
+import com.google.common.base.Predicates;
+import com.palantir.dialogue.annotations.Request;
+import com.palantir.dialogue.annotations.Request.Body;
+import com.palantir.dialogue.annotations.Request.PathParam;
+import com.palantir.logsafe.Preconditions;
+import com.squareup.javapoet.TypeName;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.MirroredTypeException;
+import javax.lang.model.type.TypeMirror;
+import org.glassfish.jersey.uri.internal.UriTemplateParser;
+
+public final class EndpointDefinitions {
+
+    private final ErrorContext errorContext;
+
+    public EndpointDefinitions(ErrorContext errorContext) {
+        this.errorContext = errorContext;
+    }
+
+    public Optional<EndpointDefinition> tryParseEndpointDefinition(ExecutableElement element) {
+        Request requestAnnotation = Preconditions.checkNotNull(element.getAnnotation(Request.class));
+        UriTemplateParser uriTemplateParser = new UriTemplateParser(requestAnnotation.path());
+
+        List<Optional<ArgumentDefinition>> args = element.getParameters().stream()
+                .map(this::getArgumentDefinition)
+                .collect(Collectors.toList());
+
+        if (!args.stream()
```

Comment:
I think I want some nice util method on ErrorContext to do this fallible construction, where normal exceptions can be thrown half way.